Refactor policy checks

eliykat opened the pull request:
## Objective

Server companion to https://github.com/bitwarden/jslib/pull/466.

This largely follows the same pattern of moving the filtering & set comparison logic into `policyService`, so that other services can just ask `policyService` "does this policy apply to this user?" without worrying about the implementation of that logic.

## Code changes
* `CurrentContext` - add `ExemptFromPolicies` method. This shadows `ManagePolicies`, which is the same criteria used in the jslib PR.
* `PolicyService` - add `PolicyAppliesToUserAsync`. It has 2 signatures for 2 use cases:
  * does a specified policy type apply to this user? (Without regard to which org the policy is for.) This is used for most policy checks, e.g. disable send, where any org policy can change what the user is allowed to do.
  * does a specified policy from a specified org apply to this user? This is used when accepting an org invite and you want to check what the policy of a specific org is.
* `Policy` model - add a helper method to deserialize the Data. This may be too simple to provide an abstraction for, but it made checking the `SendOptions.DisableHideEmail` policy more concise, and I expect we'll be using this pattern more as we add more policy options.
* `UserService` - if a user disables 2FA while being a member of orgs with 2FA policies enabled, they need to be removed. This didn't quite fit the new methods in `PolicyService`, so I at least refactored it to use similar logic, even if there's a bit of repetition here.
* all other files - refactor to use these new patterns.
* `SendServiceTests` - these have been simplified because most of the logic has been moved to `PolicyService`. Note the comment about checking `SendOptions.DisableHideEmail` - I'm open to feedback on whether there's a better way to test this.

Note for code review: please confirm I have not changed any business logic, this is intended to be a refactor only.

There is maybe another way to do this refactor: create an `OrganizationUser_ReadManyByApplicablePolicy` db procedure. That would avoid hitting various separate repositories (for policies, orgUsers and providers), avoid all the linq queries/looping, and remove the duplicate logic in `UserService`. On the other hand, that would put a lot of business logic in the db layer, which is difficult to change - so didn't end up going in that direction. Maybe that's just excessive refactoring.

I'd like feedback on this overall approach before writing additional `PolicyService` tests, in case this isn't the right solution.

eliykat commented:
`SendService` Tests are failing because I've changed dependencies. I need to look at how we're testing this now that I've moved the logic up into `PolicyService` instead.

addisonbeck commented on src/Core/Services/Implementations/PolicyService.cs:
```diff
@@ -127,5 +131,59 @@ public class PolicyService : IPolicyService
             await _eventService.LogPolicyEventAsync(policy, Enums.EventType.Policy_Updated);
         }
+
+        public async Task<bool> PolicyAppliesToCurrentUserAsync(PolicyType policyType,
```
Generally I'm not a fan of using CurrentContext in the service. I know sometimes it's necessary, but I don't think it is right now. This method has more value for reuse later if we just pass UserId in as a parameter, and determine what user that is based on what API endpoint is hit (pass CurrentContext.UserId.Value from the controller).

addisonbeck commented on src/Core/Services/Implementations/PolicyService.cs:
```diff
@@ -127,5 +131,59 @@ public class PolicyService : IPolicyService
+                return false;
+            }
+
+            var checkExemptionTasks = orgUsers
+                .Where(ou =>
+                    ou.Status >= OrganizationUserStatusType.Accepted &&
+                    enabledPolicies.Contains(ou.OrganizationId))
+                .Select(ou => _currentContext.ExemptFromPolicies(ou.OrganizationId))
+                .ToArray();
```
This would also need to change from a call to _currentContext to an explicit check using the org user records you already pulled and targeting the userId that is passed in to the function.

addisonbeck commented on src/Core/Services/Implementations/PolicyService.cs:
```diff
@@ -127,5 +131,59 @@ public class PolicyService : IPolicyService
+            return policy != null &&
+                policy.Enabled &&
+                orgUser != null &&
+                !await _currentContext.ExemptFromPolicies(organizationId) &&
```
Same here

addisonbeck reviewed (commented):
> There is maybe another way to do this refactor: create an OrganizationUser_ReadManyByApplicablePolicy db procedure. That would avoid hitting various separate repositories (for policies, orgUsers and providers), avoid all the linq queries/looping, and remove the duplicate logic in UserService. On the other hand, that would put a lot of business logic in the db layer, which is difficult to change - so didn't end up going in that direction. Maybe that's just excessive refactoring.

I think this is the direction we should go in, and don't think there is too much business logic involved. We have a concept for "Some people don't have more control over policies" in our database tables with Roles and more specifically Permissions.ManagePolicies. The database isn't unknown to these concepts and using them to analyze and return data seems pretty applicable to the DB to me. 

I think this might feel like a weird procedure because Roles aren't very indicative of what a user can and can't do, but that's more an issue with the way we handle roles than it is with this theoretical procedure that would make total sense if ManagePolicies was the only flag involved. 

I'll leave this open for input from you Thomas, and from the rest of @bitwarden/dept-engineering, as there is a lot of nuance involved. 

I left some other comments too. Lots to discuss here, and great work. 🙂

eliykat commented on src/Core/Models/Table/Policy.cs:
```diff
@@ -18,5 +19,14 @@ public void SetNewId()
             Id = CoreHelpers.GenerateComb();
         }
+
+        public T getDataModel<T>()
+        {
+            if (Data == null)
+            {
+                return default(T);
+            }
+            return JsonConvert.DeserializeObject<T>(Data);
+        }
```
Good call, I didn't notice the existing helper.

addisonbeck commented:
> ReadManyByApplicablePolicy

While not nearly as nice as competitor options, SQL Server supports parsing JSON with [several functions](https://docs.microsoft.com/en-us/sql/t-sql/functions/json-value-transact-sql?view=sql-server-ver15). We do this kind of thing already in some database operations like [CipherDetails](https://github.com/bitwarden/server/blob/2054e5a9263f6be9bcfc43020a2603cec6ae1a10/src/Sql/dbo/Functions/CipherDetails.sql). I was forced to reckon with JSON in SQL during EF work, so let me know if I can help with this beyond sharing links. I think this will cover your concerns around Permissions and Send Data, if not let me know about that too.

> Provider checks

We should updated ManagePolicies in jslib, I was unaware of this Provider check before, sorry for missing that in the jslib PR.

eliykat commented:
@addisonbeck I've made some progress with this, could you have a look at what I've done so far?

* new `OrganizationUserRepository` method: `GetManyByApplicablePolicyTypeAsync` (as discussed above)
* new `PolicyRepository` method: `GetManyByTypeApplicableToUserIdAsync`. This is identical to the above method, except it returns the `Policy` objects instead - for cases where you want to work with the applicable policies rather than the affected orgUser. My thinking: this is more useful than writing a db query just for `SendOptionsPolicy.Data.DisableHideEmail`, or doing tricky business to try to make such a db query more high level.
* in the SQL, new table-valued function so that the "policy applies to user" logic is all in one place; the above repository methods can just call simple `SELECT` queries from this table result.

eliykat commented:
@addisonbeck This is out of draft and ready for review. As discussed, I've reconceptualised it as a query for policies only. We don't need the `orgUserId` or anything about the `orgUser` in any of the current use cases for this. We only need the following:
* the `policy` object to inspect or operate on further (e.g. checking any options enabled under `policy.Data`)
* the `organizationId`, which we can get from the `policy` object
* whether any policies apply to the user (returned as a simple `COUNT` if that's the only info we need)
* if we do need to do anything with the `user`, we can generally use their `userId` which we already have to be able to run this query to begin with.

Keeping this focused on policies is much cleaner to think about and to implement at this time.

addisonbeck reviewed (dismissed):
This looks great, Thomas. The EF repositories still need to be implemented, but I'm fine with merging this if you are planning to handle those in a separate PR, this one has been through a lot already.

eliykat commented:
It's alright, let's make sure everything is covered off before we merge. I did forget about EF as I was iterating through the SQL side of things, but I've done them now and pushed them for review.

addisonbeck commented:
Leaving this as is for now, after a Slack discussion I believe @eliykat is adding some unit tests for EF.

eliykat commented:
@addisonbeck I've done the following:
* added EF tests. Pain points (any suggests are welcome):
  * there are lots of dependencies due to FK constraints, but I think this is unavoidable when testing on live dbs
  * I don't like having lots of unnamed inline values like this, but we don't currently have a better solution in our test projects
  * I also don't like having to manually create all the lists of repos at the start of the test method, but it seemed worth it to be able to include the SQL repo in the main loop.
* refactored the `PolicyApplicableToUser` SQL query, because using `ProviderUserProviderOrganizationDetailsView` was really overkill for what we need to do here. e.g. it involves joins with `Provider` and `Organization` which are not required.

addisonbeck reviewed (approved):
Great job once again, @eliykat
